src/spsc_queue.ml

```diff
@@ -36,29 +37,61 @@ let create ~size_exponent =
     head = Atomic.make 0;
     tail = Atomic.make 0;
     mask = size - 1;
-    array = Array.init size (fun _ -> None);
+    array = Array.make size None;
+    lock = Atomic.make None;
   }
 
-let push { array; head; tail; mask; _ } element =
+let unlock lock =
+  if Atomic.get lock <> None then
+    match Atomic.exchange lock None with
+    | Some release -> release ()
+    | None -> ()
+
+let try_push { array; head; tail; mask; lock } element =
   let size = mask + 1 in
-  let head_val = Atomic.get head in
   let tail_val = Atomic.get tail in
-  if head_val + size == tail_val then raise Full
+  let head_val = Atomic.get head in
+  if head_val + size = tail_val then false
   else (
     Array.set array (tail_val land mask) (Some element);
-    Atomic.set tail (tail_val + 1))
+    Atomic.set tail (tail_val + 1);
+    unlock lock;
+    true)
 
-let pop { array; head; tail; mask; _ } =
+let try_pop { array; head; tail; mask; lock } =
   let head_val = Atomic.get head in
   let tail_val = Atomic.get tail in
-  if head_val == tail_val then None
+  if head_val = tail_val then None
   else
     let index = head_val land mask in
     let v = Array.get array index in
     (* allow gc to collect it *)
     Array.set array index None;
     Atomic.set head (head_val + 1);
     assert (Option.is_some v);
+    unlock lock;
     v
 
 let size { head; tail; _ } = Atomic.get tail - Atomic.get head
+
+let wait ~rdv t expected_size =
+  let release, wait = rdv () in
+  let some_release = Some release in
+  if Atomic.compare_and_set t.lock None some_release then
+    if size t = expected_size then wait ()
+    else
+      let (_ : bool) = Atomic.compare_and_set t.lock some_release None in
```

Comment:
Hmm... I wonder whether this can cause problems?  It is possible that another party either manages to change the size of the queue (before the first CAS) or call `release` (between the two CASes).

Related to this, to avoid capturing the stack unnecessarily (before the waiters are added), the kcas blocking draft uses [three states](https://github.com/ocaml-multicore/kcas/blob/1d03f71bba77b920b38049f4f78db62e4ad4f36a/src/kcas.ml#L40): `Init | Resumed | Waiting of ...`.  Then both [suspend](https://github.com/ocaml-multicore/kcas/blob/1d03f71bba77b920b38049f4f78db62e4ad4f36a/src/kcas.ml#L638-L651) and [resume](https://github.com/ocaml-multicore/kcas/blob/1d03f71bba77b920b38049f4f78db62e4ad4f36a/src/kcas.ml#L547-L549) take the race condition (where resume happens before suspend is fully ready) into account.

Reply:
I guess SPSC makes things a bit simpler here as well.

Reply:
Yes it's possible for another party to call `release` even though `wait ()` will not be called... I think it's fine? (however, we do need to be careful when calling `wait ()` otherwise it's possible for the queue to be unstuck just before we install our lock, followed by no more operations to wake us up)

(I believe the `size t = expected_size` is doing something similar to your three states to determine the `Resumed` state implicitly for the queue internals)

Reply:
(To be clear, I totally agree that calling `release` when `wait` might not be called feels dirty when compared to your two-phases commit.. but I don't have an intuition for why this might break custom schedulers assumptions)